src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributes.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class ReadOnlyTagHelperAttributes<TAttribute> : IReadOnlyList<TAttribute>
+        where TAttribute : IReadOnlyTagHelperAttribute
+    {
+        protected ReadOnlyTagHelperAttributes()
+        {
+            Attributes = new List<TAttribute>();
+        }
+
+        public ReadOnlyTagHelperAttributes([NotNull] IEnumerable<TAttribute> attributes)
+        {
+            Attributes = new List<TAttribute>(attributes);
+        }
+
+        protected List<TAttribute> Attributes { get; }
+
+        public TAttribute this[int index]
+        {
+            get
+            {
+                return Attributes[index];
+            }
+        }
+
+        public TAttribute this[[NotNull] string key]
+        {
+            get
+            {
+                return Attributes.FirstOrDefault(attribute => KeyEquals(key, attribute));
+            }
+        }
+
+        public int Count
+        {
+            get
+            {
+                return Attributes.Count;
+            }
+        }
+
+        public virtual bool IsReadOnly
+        {
+            get
+            {
+                return true;
+            }
+        }
+
+        public bool Contains([NotNull] TAttribute item)
+        {
+            return Attributes.Contains(item);
+        }
+
+        public bool ContainsKey([NotNull] string key)
+        {
+            return Attributes.Any(attribute => KeyEquals(key, attribute));
+        }
+
+        public int IndexOf([NotNull] TAttribute item)
+        {
+            return Attributes.IndexOf(item);
+        }
+
+        public bool TryGetValue([NotNull] string key, out TAttribute value)
+        {
+            value = Attributes.FirstOrDefault(attribute => KeyEquals(key, attribute));
+
+            return value != null;
+        }
+
+        public bool TryGetValues([NotNull] string key, out IEnumerable<TAttribute> values)
+        {
+            values = Attributes.Where(attribute => KeyEquals(key, attribute));
+
+            return values.Any();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        public IEnumerator<TAttribute> GetEnumerator()
+        {
+            return Attributes.GetEnumerator();
+        }
+
+        internal static bool KeyEquals(string key, [NotNull] TAttribute attribute)
```

Comment:
Not sure what you mean by "providing protected constructors". I modified this to be `internal` from `protected` because it didn't have value being accessible to users.